src/librustc/ty/layout.rs

```diff
@@ -2648,12 +2648,11 @@ where
         };
 
         let target = &cx.tcx().sess.target.target;
-        let win_x64_gnu =
-            target.target_os == "windows" && target.arch == "x86_64" && target.target_env == "gnu";
-        let linux_s390x =
-            target.target_os == "linux" && target.arch == "s390x" && target.target_env == "gnu";
-        let linux_sparc64 =
-            target.target_os == "linux" && target.arch == "sparc64" && target.target_env == "gnu";
+        let indirect_zst = match target.arch.as_ref() {
+            "powerpc" | "s390x" | "sparc64" => true,
```

Comment:
Wait, you've removed the `target_env == "gnu"` - this workaround *only* makes sense in that case, because ZSTs in C are a GNU extension.

Reply:
@eddyb That's not true. It makes equally as much sense for `target_env == "musl"`, which follows the exact same ELF ABI as glibc. As I mentioned before:
> With these patches, I get a clean testsuite run on `powerpc-unknown-linux-musl`.
It is also necessary for the BSDs, since again they use the same ABI (both as specified, and as implemented in gcc/clang).

Reply:
You could probably add a `has_gnu_extensions` or something that's set for `target_env` being either `"gnu"` or `"musl"`.

While this doesn't feel ideal, I'd certainly prefer to broaden the whitelist than start including combinations we might not be aware of, and it does make sense that `musl` would hit the same problems.

Reply:
That still doesn't make any sense to me. `target_env` refers to the libc in use: glibc versus musl. That has nothing to do with the ABI, which is a property of the compiler.

Like I already said, this fix applies for the BSDs as well as Linux (regardless of `target_env`), because they all use GCC or a GCC-compatible compiler.